routes/webhook/gitlab.go

```diff
@@ -0,0 +1,451 @@
+/*
+
+Copyright (C) 2017-2019  Daniele Rondina <geaaru@sabayonlinux.org>
+                         Ettore Di Giacinto <mudler@gentoo.org>
+Credits goes also to Gogs authors, some code portions and re-implemented design
+are also coming from the Gogs project, which is using the go-macaron framework
+and was really source of ispiration. Kudos to them!
+
+This program is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+*/
+
+package webhook
+
+import (
+	"fmt"
+	"net/http"
+	"strconv"
+
+	"github.com/MottainaiCI/mottainai-server/pkg/context"
+	database "github.com/MottainaiCI/mottainai-server/pkg/db"
+	logging "github.com/MottainaiCI/mottainai-server/pkg/logging"
+	mottainai "github.com/MottainaiCI/mottainai-server/pkg/mottainai"
+	setting "github.com/MottainaiCI/mottainai-server/pkg/settings"
+	tasks "github.com/MottainaiCI/mottainai-server/pkg/tasks"
+	user "github.com/MottainaiCI/mottainai-server/pkg/user"
+	mhook "github.com/MottainaiCI/mottainai-server/pkg/webhook"
+
+	logrus "github.com/sirupsen/logrus"
+	webhooks "gopkg.in/go-playground/webhooks.v3"
+	gitlab "gopkg.in/go-playground/webhooks.v3/gitlab"
+)
+
+type GitLabWebhookDataResp struct {
+	TaskId     string `json:"task_id,omitempty"`
+	PipelineId string `json:"pipeline_id,omitempty"`
+	Note       string `json:"note,omitempty"`
+	Processed  bool   `json:"processed"`
+	Status     bool   `json:"status"`
+}
+
+type GitLabWebHook struct {
+	*GitWebHook
+}
+
+func NewGitLabWebHook(payload interface{}, w *mhook.WebHook, u *user.User, header webhooks.Header, kindEvent string) *GitLabWebHook {
```

Comment:
Nitpick: Can't you make it return the `WebHookCallbacks` interface at this point? you would enforce a bit more the usage in all the code

Reply:
I thought of this but the main issue is the limitation of the v3 library. I saw that with v5 we could rewrite better the code to avoid code duplication. So probably, we could postpone this with v5 API migration. WDYT?